source/openqasm/parser/antlr/qasm_parser.py

```diff
@@ -0,0 +1,926 @@
+    """Get the span of a node"""
+    if isinstance(node, ParserRuleContext):
+        return Span(node.start.line, node.start.column, node.stop.line, node.stop.column)
+    else:
+        return Span(node.symbol.line, node.symbol.start, node.symbol.line, node.symbol.stop)
+
+
+def add_span(open_node: OpenNode, span: Span) -> OpenNode:
+    """Set the span of a node and return the node"""
+    open_node.span = span
+    return open_node
+
+
+def combine_span(first: Span, second: Span):
+    """Combine two spans and return the combined one"""
+    return Span(first.start_line, first.start_column, second.start_line, second.start_column)
+
+
+def span(func):
+    """Function decorator to automatic attach span to nodes for visit* methods."""
+
+    def wrapped(*args, **kwargs):
+        span = get_span(args[1])  # args[1] is ctx
+        node = func(*args, **kwargs)
+        if node is None:
+            raise ValueError(f"None encountered at {span}")
+        else:
+            node.span = span
+            return node
+
+    return wrapped
+
+
+class OpenNodeVisitor(qasm3Visitor):
+    @span
+    def visitProgram(self, ctx: qasm3Parser.ProgramContext):
+
+        version = {"major": 3, "minor": 0}  # Default is version 3.0
+
+        if ctx.header() and ctx.header().version():
+            if ctx.header().version().Integer():
+                version_int = ctx.header().version().Integer()
+                version["major"] = int(version_int.getText())
+            elif ctx.header().version().RealNumber():
+                version_real = ctx.header().version().RealNumber()
+                for i, key in enumerate(version):
+                    version[key] = int(version_real.getText().split(".")[i])
+
+            if version["major"] not in (2, 3):
+                warnings.warn(
+                    "\n OpenNode supports OpenQASM 2/3 only.\
+                               Input file is of version {}.{}.".format(
+                        *version.values()
+                    )
+                )
+
+        io_variables = []
+        if ctx.header() and ctx.header().io():
+            io_list = ctx.header().io()
+            for io in io_list:
+                identifier = add_span(
+                    Identifier(io.Identifier().getText()), get_span(io.Identifier())
+                )
+                ctype = self.visit(io.classicalType())
+                io_variables.append(
+                    IODeclaration(
+                        io_identifier=IOIdentifierName[io.getChild(0).getText()],
+                        type=ctype,
+                        identifier=identifier,
+                        init_expression=None,
+                    )
+                )
+
+        statements = []
+        for globalStatement in ctx.globalStatement():
+            statements.append(self.visit(globalStatement))
+        for statement in ctx.statement():
+            statements.append(self.visit(statement))
+        program = Program(statements=statements)
```

Comment:
can we sort `statements` with `span.start_line` and `span.column` before constructing `Program`?

Reply:
Good catch! Added sorting.